feat: Added diff() function to DataFrame

NeonSpork opened the pull request:
Added a diff() function to mimic the functionality of the equivalent function in pandas as best as possible.

Can calculate difference between rows and columns both forwards and backwards, in addition to difference in relation to another dataframe or series.

Added the function to the DataFrame interface.

Added unit testing with as much coverage as I could. The error that is thrown if the axis is set as something other than 0 or 1 follows the convension from a similar function, but as far as I can tell TypeScript won't let you set axis to anything besides 0 or 1, so it is untestable.

The implementation is made using the built in tensorflow functionality for pretty much everything except nudging the data to compare, and I believe this is one of the faster ways to achieve this!

NeonSpork commented:
Resolves #383 partially, there's no `pct_change` as of yet, but figuring out how to perform the calculations effectively and manipulate the dataframes took the most amount of time. I will try to implement a similar function for `pct_change()` as well 😃

risenW commented on src/danfojs-base/core/frame.ts:
```diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+        }
+
+        if (typeof other === "number" && axis === 1) {
+            const orig_tensor = this.tensor.clone();
```
Hey, @NeonSpork I guess you missed it, but we use `camelCase` styling for variable names.  So instead of `orig_tensor` you use `originalTensor`

NeonSpork replied:
I'll fix that right away! Sorry about that

risenW commented on src/danfojs-base/core/frame.ts:
````diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+     * console.log(df4)
+     * ```
+    */
+    diff(other: DataFrame | Series | number[] | number, options?: { axis?: 0 | 1, inplace?: boolean }): DataFrame
````
Just to clarify are you modeling this after the pandas diff it the numpy diff function?

NeonSpork replied:
I was trying to emulate the pandas, but it was very difficult to have the default of `diff()` does difference with previous row. So I erred on the side of caution and made it such that the user must explicitly state what they want to calculate the difference with!

risenW replied:
Good point.

risenW commented on src/danfojs-base/core/frame.ts:
```diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+        }
+
+        if (typeof other === "number" && axis === 0) {
+            const orig_df = new DataFrame(this.tensor.clone());
```
You can easily deep copy a DataFrame by calling `df.copy()`

NeonSpork replied:
This is a great point! I'll try to refactor!

risenW commented on src/danfojs-base/core/frame.ts:
```diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+            const orig_df = new DataFrame(this.tensor.clone());
+            const orig_tensor = orig_df.T.tensor.clone();
+            let unit = [NaN];
+            for (let i = 1; i < orig_tensor.shape[orig_tensor.rank - 1]; i++) {
```
**Suggestion**: If you need to create an array of a specific size with default values, you can use:
```javascript
new Array(orig_tensor.rank - 1).fill(NaN)
```

NeonSpork replied:
Ah, that's a faster/better way to do it

NeonSpork replied:
Had to grab the final number in the shape to get the correct element, but that was a much shorter and simpler way to do it!
```typescript
const unit = new Array(originalTensor.shape[originalTensor.rank - 1]).fill(NaN);
```

risenW commented on src/danfojs-base/core/frame.ts:
```diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+            }
+            const diff_tensor = tensorflow.tensor2d(diff_array, orig_tensor.shape);
+            const diff_df = this.$MathOps([orig_tensor, diff_tensor], "sub", inplace) as DataFrame;
+            return diff_df.T;
```
Suggestion: If this is the only difference between when axis=1 or 0, then a cleaner method will be to perform the computation and at the end based on the axis you can return the transpose:

```javascript
// perform all computation ...
// ...

if (axis == 0) {
  return diff_df
} else {
  return diff_df.T;
}



```

NeonSpork replied:
The problem with doing this is that I'm essentially flipping the entire thing with `df.T` first and then pretending the columns are rows, then flipping them back at the end.

I'll try to refactor it so it's more succinct and less duplicated!

risenW reviewed (commented):
Great work so far @NeonSpork ! and thanks for taking out time to contribute. I dropped some general suggestions, if you need any input from me, don't hesitate to ping mw.

NeonSpork commented:
Another comment. I have implemented the pct_change() as well, but am struggling with testing.

My expected return is for example `0.8` but JavaScript thinks it should be `0.80000000000012345` to provide a random example. Is there an easy way around this?

risenW commented:
> Another comment. I have implemented the pct_change() as well, but am struggling with testing.
> 
> My expected return is for example `0.8` but JavaScript thinks it should be `0.80000000000012345` to provide a random example. Is there an easy way around this? @risenW
> 
> Edit: Since it breaks testing I'm doing `pct_change()` in a separate branch. I can make a separate pull request for that in a bit, so you can see more hands on!

Yeah, I have had that issue with float values as well. You can try to truncate/floor the value. 

And yes, doing pct_change in a separate PR is better.

NeonSpork commented:
> 

We can cross that bridge in another PR! 👍 I'm trying to refactor and remove some duplication, should hopefully not take too long!

NeonSpork commented:
Alright! I think I've resolved the issues you pointed out @risenW ?

The code has become more succinct with less duplication, and tests are passing!

risenW commented:
@NeonSpork Everything looks good here. One last thing before I merge, can you also add test for the browser version. Just copy over what you currently have to the browser test folder, and then use the namespace `dfd.DataFrame` or `dfd.Series` to create Dataframe and Series. You can see examples in the `danfojs-browser/tests/` folder

NeonSpork commented:
Sounds great! I forgot about the browser tests, oops! Was just happy to see green checks 👍

I'll sort that out in both my PRs in a short bit!

risenW commented:
LGTM ✅✅ Thanks for the PR @NeonSpork

NeonSpork commented:
> LGTM ✅✅ Thanks for the awesome work @NeonSpork

My pleasure man! I'm looking forward to contributing more in the future 🚀

NeonSpork commented:
> @NeonSpork @risenW - this feature is exactly what I am searching for, however, I cannot find it in the docs. Could you maybe point me to the documentation?

Hi! Glad you've found my contribution useful! When I wrote that contribution I also added pages to the docs, but for some reason they're not on the official documentation web page. Not sure why that is the case.

Either way the docs pages I added are still in the docs repo, and the examples and explanations for the `diff()` function can be found here:
https://github.com/javascriptdata/danfojs-doc/blob/master/api-reference%2Fdataframe%2Fdanfo.dataframe.diff.md

Maybe @risenW has some insight into why it's not on the main docs? Maybe the main index page doesn't include links to them?
